File: Sources/DebridStreamer/Services/Sync/TraktSyncService.swift
import Foundation

enum SyncState: String, Sendable, Codable {
    case idle
    case running
    case success
    case failed
}

enum TraktSyncError: LocalizedError, Equatable {
    case invalidURL
    case invalidResponse
    case decodingFailed(String)
    case httpStatus(Int, String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid Trakt URL."
        case .invalidResponse:
            return "Invalid Trakt response."
        case .decodingFailed(let detail):
            return "Failed to decode Trakt response: \(detail)"
        case .httpStatus(let status, let body):
            return "Trakt HTTP \(status): \(body)"
        }
    }
}

struct TraktDeviceCodeResponse: Codable, Sendable {
    var deviceCode: String
    var userCode: String
    var verificationURL: String
    var expiresIn: Int
    var interval: Int

    enum CodingKeys: String, CodingKey {
        case deviceCode = "device_code"
        case userCode = "user_code"
        case verificationURL = "verification_url"
        case expiresIn = "expires_in"
        case interval
    }
}

struct TraktTokenResponse: Codable, Sendable {
    var accessToken: String
    var refreshToken: String
    var expiresIn: Int
    var tokenType: String
    var scope: String
    var createdAt: Int

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresIn = "expires_in"
        case tokenType = "token_type"
        case scope
        case createdAt = "created_at"
    }
}

struct TraktWatchlistItem: Sendable, Equatable {
    var imdbID: String
    var title: String
    var year: Int?
}

/// Typed summary of a `POST /sync/watchlist` response. Trakt returns counts of
/// items that were `added`, were already `existing`, and could `not_found` be
/// matched. Decoding this (instead of discarding it) lets callers surface
/// rejected IDs to the user.
struct TraktWatchlistPushResult: Decodable, Sendable, Equatable {
    struct Counts: Decodable, Sendable, Equatable {
        var movies: Int?
    }

    struct NotFoundIDs: Decodable, Sendable, Equatable {
        var imdb: String?
    }

    struct NotFoundMovie: Decodable, Sendable, Equatable {
        var ids: NotFoundIDs?
    }

    struct NotFound: Decodable, Sendable, Equatable {
        var movies: [NotFoundMovie]?
    }

    var added: Counts?
    var existing: Counts?
    var notFound: NotFound?

    enum CodingKeys: String, CodingKey {
        case added
        case existing
        case notFound = "not_found"
    }
}

/// Typed summary of a `POST /scrobble/{action}` response. Trakt echoes the
/// reported `progress` and an `action` (e.g. `start`, `pause`, `scrobble`). All
/// fields are optional so a thin/best-effort response still decodes.
struct TraktScrobbleResult: Decodable, Sendable, Equatable {
    var action: String?
    var progress: Double?
}

actor TraktSyncService {
    private let session: URLSession
    private let baseURL = "https://api.trakt.tv"

    /// Default safety buffer (seconds) before the real expiry at which a token is
    /// considered expired, so callers refresh proactively rather than racing a 401.
    static let defaultExpiryBuffer: TimeInterval = 24 * 60 * 60

    init(session: URLSession = .shared) {
        self.session = session
    }

    /// Returns true when a token issued at `createdAt` (Unix seconds) with lifetime
    /// `expiresIn` (seconds) is at or past its expiry, accounting for `buffer`.
    /// Both `createdAt` and `expiresIn` come directly from `TraktTokenResponse`.
    nonisolated static func isExpired(
        createdAt: Int,
        expiresIn: Int,
        now: Date = Date(),
        buffer: TimeInterval = TraktSyncService.defaultExpiryBuffer
    ) -> Bool {
        let expiry = Date(timeIntervalSince1970: TimeInterval(createdAt + expiresIn))
        return expiry.timeIntervalSince(now) <= buffer
    }

    func startDeviceAuth(clientID: String) async throws -> TraktDeviceCodeResponse {
        let payload = ["client_id": clientID]
        return try await request(
            path: "/oauth/device/code",
            method: "POST",
            traktClientID: nil,
            accessToken: nil,
            body: payload
        )
    }

    func exchangeDeviceCode(
        clientID: String,
        clientSecret: String,
        deviceCode: String
    ) async throws -> TraktTokenResponse {
        let payload: [String: String] = [
            "code": deviceCode,
            "client_id": clientID,
            "client_secret": clientSecret
        ]
        return try await request(
            path: "/oauth/device/token",
            method: "POST",
            traktClientID: nil,
            accessToken: nil,
            body: payload
        )
    }

    func refreshToken(
        clientID: String,
        clientSecret: String,
        refreshToken: String
    ) async throws -> TraktTokenResponse {
        let payload: [String: String] = [
            "refresh_token": refreshToken,
            "client_id": clientID,
            "client_secret": clientSecret,
            "grant_type": "refresh_token"
        ]
        return try await request(
            path: "/oauth/token",
            method: "POST",
            traktClientID: nil,
            accessToken: nil,
            body: payload
        )
    }

    func fetchWatchlist(clientID: String, accessToken: String) async throws -> [TraktWatchlistItem] {
        struct ResponseItem: Decodable {
            struct Movie: Decodable {
                struct IDs: Decodable {
                    let imdb: String?
                }
                let title: String
                let year: Int?
                let ids: IDs
            }
            let movie: Movie?
        }

        let response: [ResponseItem] = try await request(
            path: "/sync/watchlist/movies",
            method: "GET",
            traktClientID: clientID,
            accessToken: accessToken,
            body: Optional<String>.none
        )

        return response.compactMap { item in
            guard let movie = item.movie, let imdb = movie.ids.imdb else { return nil }
            return TraktWatchlistItem(imdbID: imdb, title: movie.title, year: movie.year)
        }
    }

    @discardableResult
    func pushWatchlist(clientID: String, accessToken: String, imdbIDs: [String]) async throws -> TraktWatchlistPushResult {
        struct Payload: Encodable {
            struct Movie: Encodable {
                struct IDs: Encodable {
                    let imdb: String
                }
                let ids: IDs
            }
            let movies: [Movie]
        }

        let payload = Payload(movies: imdbIDs.map { .init(ids: .init(imdb: $0)) })
        return try await request(
            path: "/sync/watchlist",
            method: "POST",
            traktClientID: clientID,
            accessToken: accessToken,
            body: payload
        )
    }

    /// The Trakt scrobble action. `start` marks playback began, `pause` a pause,
    /// and `stop` finalizes - Trakt auto-marks the item watched when a `stop`
    /// arrives with progress >= 80%.
    enum ScrobbleAction: String, Sendable {
        case start
        case pause
        case stop
    }

    /// Scrobbles a movie's playback progress to Trakt. `progressPercent` is the
    /// 0...100 watched fraction. Only IMDb-identified movies are supported (the
    /// id must look like `tt…`); callers should pass the bare IMDb id.
    ///
    /// This is best-effort: callers invoke it fire-and-forget and ignore errors so
    /// playback is never blocked or interrupted by a Trakt failure.
    @discardableResult
    func scrobbleMovie(
        clientID: String,
        accessToken: String,
        imdbID: String,
        progressPercent: Double,
        action: ScrobbleAction
    ) async throws -> TraktScrobbleResult {
        struct Payload: Encodable {
            struct Movie: Encodable {
                struct IDs: Encodable { let imdb: String }
                let ids: IDs
            }
            let movie: Movie
            let progress: Double
        }

        let clamped = min(max(progressPercent, 0), 100)
        let payload = Payload(movie: .init(ids: .init(imdb: imdbID)), progress: clamped)
        return try await request(
            path: "/scrobble/\(action.rawValue)",
            method: "POST",
            traktClientID: clientID,
            accessToken: accessToken,
            body: payload
        )
    }

    /// Scrobbles an episode's playback progress to Trakt, identifying the show by
    /// IMDb id plus season/episode numbers.
    @discardableResult
    func scrobbleEpisode(
        clientID: String,
        accessToken: String,
        showIMDbID: String,
        season: Int,
        episode: Int,
        progressPercent: Double,
        action: ScrobbleAction
    ) async throws -> TraktScrobbleResult {
        struct Payload: Encodable {
            struct Show: Encodable {
                struct IDs: Encodable { let imdb: String }
                let ids: IDs
            }
            struct Episode: Encodable {
                let season: Int
                let number: Int
            }
            let show: Show
            let episode: Episode
            let progress: Double
        }

        let clamped = min(max(progressPercent, 0), 100)
        let payload = Payload(
            show: .init(ids: .init(imdb: showIMDbID)),
            episode: .init(season: season, number: episode),
            progress: clamped
        )
        return try await request(
            path: "/scrobble/\(action.rawValue)",
            method: "POST",
            traktClientID: clientID,
            accessToken: accessToken,
            body: payload
        )
    }

    private func request<T: Decodable, Body: Encodable>(
        path: String,
        method: String,
        traktClientID: String?,
        accessToken: String?,
        body: Body?
    ) async throws -> T {
        guard let url = URL(string: baseURL + path) else {
            throw TraktSyncError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 45
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("2", forHTTPHeaderField: "trakt-api-version")
        if let traktClientID {
            request.setValue(traktClientID, forHTTPHeaderField: "trakt-api-key")
        }
        if let accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.httpBody = try JSONEncoder().encode(body)
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw TraktSyncError.invalidResponse
        }

        if !(200...299).contains(http.statusCode) {
            throw TraktSyncError.httpStatus(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }

        if T.self == TraktNoContent.self {
            return TraktNoContent() as! T
        }

        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw TraktSyncError.decodingFailed(String(describing: error))
        }
    }
}

private struct TraktNoContent: Decodable {}

File: web/src/services/sync/TraktSyncService.ts
// Port of Sources/DebridStreamer/Services/Sync/TraktSyncService.swift.
//
// A fetch-based Trakt sync client. Mirrors the Swift actor's behavior: the same
// API paths/methods, the same request headers (Content-Type, trakt-api-version,
// trakt-api-key, Authorization Bearer), the token expiry math, and the
// decode/error-mapping (httpStatus vs decodingFailed). The `fetch`
// implementation is injectable so tests can stub the network (the Swift code
// injects a URLSession instead). There is no caching in the Swift service, so
// - unlike the TMDB template - there is no TTL cache here.
//
// Note on decodingFailed: Swift's JSONDecoder throws when a *structurally*
// invalid body (valid JSON, wrong shape - e.g. `{"unexpected":true}`) is fed to
// a Decodable with required keys. `JSON.parse` does NOT throw for that, so each
// strongly-typed decode runs an explicit shape `decode` callback; any throw
// inside it (parse failure OR missing required field) is mapped to
// `decodingFailed`, exactly as Swift maps any decode error.

import {
  type TraktDeviceCodeResponse,
  type TraktScrobbleItem,
  type TraktTokenResponse,
  type TraktWatchlistItem,
  type TraktWatchlistShowItem,
  type TraktWatchlistPushResult,
} from "./models";
import {
  decodeDeviceCodeResponse,
  decodeScrobbleResult,
  decodeTokenResponse,
  decodeWatchlistItems,
  decodeWatchlistShowItems,
  decodeWatchlistPushResult,
  encodeScrobbleItem,
  TraktSyncError,
} from "./types";

/** Injectable fetch signature (a subset of the DOM `fetch`). Mirrors the
 * `URLSession` the Swift actor injects. */
export type FetchImpl = (
  url: string,
  init?: {
    method?: string;
    headers?: Record<string, string>;
    body?: string;
  },
) => Promise<{
  status: number;
  text(): Promise<string>;
}>;

export class TraktSyncService {
  private readonly baseURL = "https://api.trakt.tv";
  private readonly fetchImpl: FetchImpl;

  /** Default safety buffer (seconds) before the real expiry at which a token is
   * considered expired, so callers refresh proactively rather than racing a
   * 401. Mirrors Swift `defaultExpiryBuffer` (24h). */
  static readonly defaultExpiryBuffer = 24 * 60 * 60;

  constructor(fetchImpl?: FetchImpl) {
    // Default to the global fetch; tests inject a stub.
    this.fetchImpl =
      fetchImpl ?? ((url, init) => fetch(url, init as RequestInit));
  }

  /**
   * Returns true when a token issued at `createdAt` (Unix seconds) with lifetime
   * `expiresIn` (seconds) is at or past its expiry, accounting for `buffer`.
   * Both `createdAt` and `expiresIn` come directly from `TraktTokenResponse`.
   * `now` is epoch seconds (Swift passes a `Date`; here a number keeps it pure).
   * Mirrors Swift `isExpired`.
   */
  static isExpired(
    createdAt: number,
    expiresIn: number,
    now: number = Date.now() / 1000,
    buffer: number = TraktSyncService.defaultExpiryBuffer,
  ): boolean {
    const expiry = createdAt + expiresIn;
    return expiry - now <= buffer;
  }

  // MARK: - OAuth device flow

  async startDeviceAuth(clientID: string): Promise<TraktDeviceCodeResponse> {
    return this.request({
      path: "/oauth/device/code",
      method: "POST",
      body: { client_id: clientID },
      decode: decodeDeviceCodeResponse,
    });
  }

  async exchangeDeviceCode(
    clientID: string,
    clientSecret: string,
    deviceCode: string,
  ): Promise<TraktTokenResponse> {
    return this.request({
      path: "/oauth/device/token",
      method: "POST",
      body: {
        code: deviceCode,
        client_id: clientID,
        client_secret: clientSecret,
      },
      decode: decodeTokenResponse,
    });
  }

  async refreshToken(
    clientID: string,
    clientSecret: string,
    refreshToken: string,
  ): Promise<TraktTokenResponse> {
    return this.request({
      path: "/oauth/token",
      method: "POST",
      body: {
        refresh_token: refreshToken,
        client_id: clientID,
        client_secret: clientSecret,
        grant_type: "refresh_token",
      },
      decode: decodeTokenResponse,
    });
  }

  // MARK: - Sync

  async fetchWatchlist(
    clientID: string,
    accessToken: string,
  ): Promise<TraktWatchlistItem[]> {
    return this.request({
      path: "/sync/watchlist/movies",
      method: "GET",
      traktClientID: clientID,
      accessToken,
      decode: decodeWatchlistItems,
    });
  }

  async fetchWatchlistShows(
    clientID: string,
    accessToken: string,
  ): Promise<TraktWatchlistShowItem[]> {
    return this.request({
      path: "/sync/watchlist/shows",
      method: "GET",
      traktClientID: clientID,
      accessToken,
      decode: decodeWatchlistShowItems,
    });
  }

  async pushWatchlist(
    clientID: string,
    accessToken: string,
    imdbIDs: string[],
    showTMDBIDs: number[] = [],
  ): Promise<TraktWatchlistPushResult> {
    const payload = {
      movies: imdbIDs.map((imdb) => ({ ids: { imdb } })),
      shows: showTMDBIDs.map((tmdb) => ({ ids: { tmdb } })),
    };
    return this.request({
      path: "/sync/watchlist",
      method: "POST",
      traktClientID: clientID,
      accessToken,
      body: payload,
      decode: decodeWatchlistPushResult,
    });
  }

  async scrobbleStart(
    clientID: string,
    accessToken: string,
    item: TraktScrobbleItem,
  ): Promise<void> {
    return this.scrobble("start", clientID, accessToken, item);
  }

  async scrobblePause(
    clientID: string,
    accessToken: string,
    item: TraktScrobbleItem,
  ): Promise<void> {
    return this.scrobble("pause", clientID, accessToken, item);
  }

  async scrobbleStop(
    clientID: string,
    accessToken: string,
    item: TraktScrobbleItem,
  ): Promise<void> {
    return this.scrobble("stop", clientID, accessToken, item);
  }

  private async scrobble(
    action: "start" | "pause" | "stop",
    clientID: string,
    accessToken: string,
    item: TraktScrobbleItem,
  ): Promise<void> {
    await this.request({
      path: `/scrobble/${action}`,
      method: "POST",
      traktClientID: clientID,
      accessToken,
      body: encodeScrobbleItem(item),
      decode: decodeScrobbleResult,
    });
  }

  // MARK: - HTTP
  //
  // Mirrors the Swift `request` generic: builds the URL, sets the fixed headers,
  // optionally sets trakt-api-key / Authorization, encodes the JSON body, then
  // maps non-2xx to `httpStatus` and a decode failure to `decodingFailed`.

  private async request<T>(opts: {
    path: string;
    method: string;
    traktClientID?: string | null;
    accessToken?: string | null;
    body?: unknown;
    decode: (raw: unknown) => T;
  }): Promise<T> {
    let url: string;
    try {
      url = new URL(this.baseURL + opts.path).toString();
    } catch {
      throw TraktSyncError.invalidURL();
    }

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "trakt-api-version": "2",
    };
    if (opts.traktClientID != null) {
      headers["trakt-api-key"] = opts.traktClientID;
    }
    if (opts.accessToken != null) {
      headers.Authorization = `Bearer ${opts.accessToken}`;
    }

    const response = await this.fetchImpl(url, {
      method: opts.method,
      headers,
      body: opts.body != null ? JSON.stringify(opts.body) : undefined,
    });

    const status = response.status;
    const text = await response.text();

    if (!(status >= 200 && status <= 299)) {
      throw TraktSyncError.httpStatus(status, text);
    }

    let parsed: unknown;
    try {
      parsed = JSON.parse(text);
    } catch (error) {
      throw TraktSyncError.decodingFailed(String(error));
    }

    try {
      return opts.decode(parsed);
    } catch (error) {
      if (error instanceof TraktSyncError) throw error;
      throw TraktSyncError.decodingFailed(String(error));
    }
  }
}
